04 - Reserving/Data Simulation/_scripts/feature_generation.py

```diff
@@ -0,0 +1,182 @@
+### inflation = growth parameters (per LoB) for the numbers of claims in the 12 accident years
+### seed1 = set the seed for reproducibility
+
+    import numpy as np
+    import pandas as pd
+    import scipy as sp
+    from scipy import stats
+    import math
+    import random
+
+# Weights in LoB.dist have to be nonnegative
+    if len([num for num in lob if num < 0]) > 0:
+        print("ERROR: The weights determining the distribution amongst the lines of business cannot be negative.")
+        features=pd.DataFrame(np.zeros((0,7)),columns=['ClNr', 'LoB', 'cc', 'AY', 'AQ', 'age', 'inj_part'])
+        return features
+
+# Preparation for the data set "features" 
+# Determine the number of claims per line of business
+    vdist = np.zeros((13,4))
+    random.seed(seed1)
+    rozkl=np.random.multinomial(V, lob, size=1)
+# Determine the number of claims per accident year (for all lines of business)
+    vdist[0]=rozkl
+    random.seed(seed1+1)
+    vmean=np.array([inflation,]*11)
+    vsd=np.array([np.abs(inflation),]*11)
+    vdist[2:13,:4]=np.random.normal(vmean,vsd)
+    vdist[2:13,:4]=np.cumsum(vdist[2:13,:4],axis=0)
+    vdist[1:13,:4]=np.exp(vdist[1:13,:4])
+    vdist_pom=vdist[1:13,:4].transpose()/vdist[1:13,:4].sum(axis=0)[:,None]
+    vdist[1:13,:4]=vdist_pom.transpose()
+    vdist_pom2=vdist[1:13,:4].transpose()*vdist[0,:4][:,None]
+    vdist[1:13,:4]=vdist_pom2.transpose()
+    random.seed(seed1+2)
+    vdist[1:13,:4]=np.random.poisson(vdist[1:13,:4],(12,4))
+    vdist[0,:4]=vdist[1:13,:4].sum(axis=0)[:,None].transpose()
+
+# Create the array where we will store the observations
+    features=pd.DataFrame(np.zeros((int(vdist[0,:4].sum()),7)),columns=['ClNr', 'LoB', 'cc', 'AY', 'AQ', 'age', 'inj_part'])
+
+# Store line of business
+    features.LoB=np.repeat(np.array([1,2,3,4]),vdist[0,:4].astype(int),axis=0).transpose()
+
+# Store accident year
+    a=np.concatenate(vdist[1:13,:4].transpose())[np.where(np.concatenate(vdist[1:13,:4].transpose())>0)]
+    b=np.where(np.concatenate(vdist[1:13,:4].transpose())>0)[0]%12
+    features.AY=(np.repeat(b,a.astype(int),axis=0)+1994).transpose()
+
+# Add artificial observations that prevent data sets below from being empty
+    features = features.append({'ClNr':-1,'LoB':1,'cc':0,'AY':1994,'AQ':0,'age':0,'inj_part':0}, ignore_index=True)
+    features = features.append({'ClNr':-1,'LoB':1,'cc':0,'AY':1994,'AQ':0,'age':0,'inj_part':0}, ignore_index=True)
+    features = features.append({'ClNr':-1,'LoB':3,'cc':0,'AY':1994,'AQ':0,'age':0,'inj_part':0}, ignore_index=True)
+    features = features.append({'ClNr':-1,'LoB':3,'cc':0,'AY':1994,'AQ':0,'age':0,'inj_part':0}, ignore_index=True)
+
+
+#feature generation for LoB 1 and 2
+
+# Generate observations from a multivariate normal distribution
+    random.seed(seed1+3)
+    Sigma=pd.DataFrame(pd.read_table('./dane/Simulation.Machine.V1/Feature.Generation.Parameters/LoB1and2/Covariances/Covariance.txt', sep='\t'))
```

Comment:
Please use the dictionary generated by the `read_project` function from `initialize.py`.